src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs

```diff
@@ -41,172 +41,217 @@ public override void Initialize(AnalysisContext context)
+            symbolStartContext.RegisterOperationBlockStartAction(OnOperationBlockStart);
 
-                // Disposable fields with initializer at declaration must be disposed.
-                compilationContext.RegisterOperationAction(operationContext =>
+            // Report diagnostics at symbol end.
+            symbolStartContext.RegisterSymbolEndAction(OnSymbolEnd);
+
+            return;
+
+            // Local functions
+            void AddOrUpdateFieldDisposedValue(IFieldSymbol field, bool disposed)
+            {
+                Debug.Assert(!field.IsStatic);
+                Debug.Assert(disposeAnalysisHelper!.IsDisposable(field.Type));
+
+                fieldDisposeValueMap.AddOrUpdate(field,
+                    addValue: disposed,
+                    updateValueFactory: (f, currentValue) => currentValue || disposed);
+            }
+
+            bool ShouldAnalyze(INamedTypeSymbol namedType)
+            {
+                // We only want to analyze types which are disposable (implement System.IDisposable directly or indirectly)
+                // and have at least one disposable field.
+                return !hasErrors &&
+                    disposeAnalysisHelper!.IsDisposable(namedType) &&
+                    !disposeAnalysisHelper.GetDisposableFields(namedType).IsEmpty &&
+                    !namedType.IsConfiguredToSkipAnalysis(symbolStartContext.Options, Rule, symbolStartContext.Compilation, symbolStartContext.CancellationToken);
+            }
+
+            bool IsDisposeMethod(IMethodSymbol method)
+                => disposeAnalysisHelper!.GetDisposeMethodKind(method) != DisposeMethodKind.None;
+
+            bool HasDisposeMethod(INamedTypeSymbol namedType)
+            {
+                foreach (var method in namedType.GetMembers().OfType<IMethodSymbol>())
                 {
-                    if (!ShouldAnalyze(operationContext.ContainingSymbol.ContainingType))
+                    if (IsDisposeMethod(method))
                     {
-                        return;
+                        return true;
                     }
+                }
 
-                    var initializedFields = ((IFieldInitializerOperation)operationContext.Operation).InitializedFields;
-                    foreach (var field in initializedFields)
-                    {
-                        if (!field.IsStatic &&
-                            disposeAnalysisHelper.GetDisposableFields(field.ContainingType).Contains(field))
-                        {
-                            addOrUpdateFieldDisposedValue(field, disposed: false);
-                        }
-                    }
-                },
-                OperationKind.FieldInitializer);
+                return false;
+            }
+
+            void OnFieldInitializer(OperationAnalysisContext operationContext)
+            {
+                RoslynDebug.Assert(disposeAnalysisHelper != null);
```

Comment:
Is there any guideline on when to use `RoslynDebug` or `Debug`?

Reply:
`RoslynDebug.Assert` is something that is nullable annotated, so we use it only when we get a false nullable warning from the C# compiler. Other alternative is to use the `!` suppression operator at use sites for this variable.

Reply:
Thanks